Fix writing of track metadata on Windows

daschuer opened the pull request:
This fixes https://bugs.launchpad.net/mixxx/+bug/1955314 and https://bugs.launchpad.net/mixxx/+bug/1955331. 

This is a mayor bugfix that should be released soon. Because the track metadata was not exported on windows when the track has not been loaded into a deck. Since the user assumes that the metadata is written, there is a risk of data loss. This was discovered in the first attempt for a fix here: https://github.com/mixxxdj/mixxx/pull/4572 

The fix makes use of the ReplaceFileW() API function that is recommended instead of the deprecated transition feature for atomic file operations. ReplaceFileW() avoids virus scanner and indexer locking and keeps the file creation time at its old value. 

The refactoring part has been issued here https://github.com/mixxxdj/mixxx/pull/4584 for the main branch.

daschuer committed:
Use dedicated  ReplaceFileW API function when relpacing the track after writing

This aims to fix the lost file creation time and the file access issue

daschuer committed:
Handle ERROR_SHARING_VIOLATION and ERROR_ACCESS_DENIED

daschuer committed:
Added a retry loop in case of a temprary ERROR_SHARING_VIOLATION

daschuer committed:
Close track file after reading stream data before metadata export

This fixes metadata export on Windows when the track not been played before.

uklotzde commented on src/sources/metadatasourcetaglib.cpp:
```diff
@@ -702,6 +714,97 @@ class SafelyWritableFile final {
+                            << "by"
+                            << m_tempFileName
+                            << "because it is used by another process";
+                    Sleep(kWindowsSharingViolationsSleepMs); // Note: Capital S for Windows API Sleep()
```
Why don't you use `QThread::msleep()` here?

uklotzde commented on src/sources/metadatasourcetaglib.cpp:
```diff
@@ -30,6 +34,14 @@ const QString kSafelyWritableTempFileSuffix = QStringLiteral("_temp");
+// After Mixxx has closed the track file, the indexer or virus scanner
+// might kick in and fail ReplaceFileW() with a sharing violation when
+// replacing the original file with the one with the updated metadata.
+const int kWindowsSharingViolationsRetries = 5;
```
kWindowsSharingViolationMaxRetries

uklotzde commented on src/sources/metadatasourcetaglib.cpp:
```diff
@@ -30,6 +34,14 @@ const QString kSafelyWritableTempFileSuffix = QStringLiteral("_temp");
+// might kick in and fail ReplaceFileW() with a sharing violation when
+// replacing the original file with the one with the updated metadata.
+const int kWindowsSharingViolationsRetries = 5;
+const int kWindowsSharingViolationsSleepMs = 100;
```
kWindowsSharingViolationSleepBeforeNextRetryMillis

uklotzde commented on src/sources/metadatasourcetaglib.cpp:
```diff
@@ -702,6 +714,97 @@ class SafelyWritableFile final {
+                    return false;
+                case ERROR_SHARING_VIOLATION:
+                    // The process cannot access the file because it is being used by another process.
+                    kLogger.critical()
```
Not critical if a retry is following.

uklotzde commented on src/sources/metadatasourcetaglib.cpp:
```diff
@@ -702,6 +714,97 @@ class SafelyWritableFile final {
+                    return false;
+                }
+            }
+            break;
```
It is difficult to recognize that this break belongs to the loop. Is it actually needed?

uklotzde commented on src/sources/metadatasourcetaglib.cpp:
```diff
@@ -30,6 +34,14 @@ const QString kSafelyWritableTempFileSuffix = QStringLiteral("_temp");
 const QString kSafelyWritableOrigFileSuffix = QStringLiteral("_orig");
 
+#ifdef __WINDOWS__
+// After Mixxx has closed the track file, the indexer or virus scanner
```
This comment belongs to the code where these constants are actually used, but not to these constants. Mentioning why ReplaceFileW() is needed here but not in the context were it is actually invoked is confusing.

uklotzde commented on src/sources/soundsourceproxy.cpp:
```diff
@@ -361,6 +361,12 @@ SoundSourceProxy::exportTrackMetadataBeforeSaving(Track* pTrack, UserSettingsPoi
                 pTrack->updateStreamInfoFromSource(
                         pAudioSource->getStreamInfo());
                 DEBUG_ASSERT(pTrack->hasStreamInfoFromSource());
+                // Make sure that the track file is closes after reading
```
multiple typos

uklotzde commented on src/sources/soundsourceproxy.cpp:
```diff
@@ -731,6 +737,7 @@ mixxx::AudioSourcePointer SoundSourceProxy::openAudioSource(
                 // before exporting metadata. In this case the caller (this class)
                 // is responsible for updating the stream info if needed.
                 if (!m_pTrack) {
+                    // TODO() It is surprising to receive here a m_pSoundSource;
```
- TODO is not a function
- "receive"?
- semicolon to end a comment?

daschuer committed:
Replace Sleep() by QThread::msleep()

daschuer committed:
Improve variable naming

daschuer committed:
Improve control flow and comments

daschuer commented:
Done

uklotzde commented on src/sources/soundsourceproxy.cpp:
```diff
@@ -731,6 +737,7 @@ mixxx::AudioSourcePointer SoundSourceProxy::openAudioSource(
                 // before exporting metadata. In this case the caller (this class)
                 // is responsible for updating the stream info if needed.
                 if (!m_pTrack) {
+                    // TODO: It is surprising that openAudioSource returns a m_pSoundSource
```
Why is it suprising?? `SoundSource` is an extension of both `AudioSource` and `MetadataSource`, basic OO design.

uklotzde replied:
This should work and fix the bug with the open file:

```
m_pAudioSource = m_pSoundSource;
return m_pAudioSource;
```

daschuer replied:
I have already fixed the underlying issue in https://github.com/mixxxdj/mixxx/pull/4584 please have a look.

uklotzde commented on src/sources/soundsourceproxy.cpp:
```diff
@@ -361,6 +361,12 @@ SoundSourceProxy::exportTrackMetadataBeforeSaving(Track* pTrack, UserSettingsPoi
                 pTrack->updateStreamInfoFromSource(
                         pAudioSource->getStreamInfo());
                 DEBUG_ASSERT(pTrack->hasStreamInfoFromSource());
+                // Make sure that the track file is closed after reading.
```
This additional code should become obsolete with the fix I just proposed.

uklotzde replied:
This is still needed:
```
proxy.closeAudioSource();
```

daschuer replied:
I would like to keep this as an hot fix solution with a minimum impact.

The rest is fixed in https://github.com/mixxxdj/mixxx/pull/4584

uklotzde replied:
I would prefer a proper fix here.

uklotzde replied:
I am not going to merge a hack with a wrong comment.

daschuer replied:
Why do you think the comment is wrong? Add least it is working.

But I don't mind to merge your ideas.
Can you issue a PR against my branch? 

Thank you.

daschuer replied:
The alternative would be to merge https://github.com/mixxxdj/mixxx/pull/4584 to this branch. This is IMHO the proper fix, but touches more code.

uklotzde replied:
I haven't reviewed #4584 yet and I cannot promise to do it soon.

uklotzde replied:
https://github.com/daschuer/mixxx/pull/72

Moreover the build failed on Linux with Clang due to a wrong preprocessor directive. Fixed.

uklotzde committed:
Delete redundant std::move() invocation

uklotzde committed:
Keep the class invariants of SoundSourceProxy valid

If the AudioSource is opened the internal pointer must be initialized
to ensure that it could be closed later!

uklotzde committed:
Fix preprocessor directive

Fixes the build on Linux with CLang.

uklotzde commented:
This important bugfix should be mentioned in the CHANGELOG.

daschuer committed:
Merge pull request #72 from uklotzde/lp1955331

lp1955331: Preserve class invariants of SoundSourceProxy

daschuer commented:
We have got the confirmation of that it is working: 
https://bugs.launchpad.net/mixxx/+bug/1955314

uklotzde reviewed (approved):
Thank you! LGTM

I'll merge to main
